File: pratap_dev/public/js/purchase_receipt_batch_entry.js
frappe.ui.form.on("Purchase Receipt", {
	setup(frm) {
		patch_pratap_batch_entry_handler(frm);
	},

	onload(frm) {
		patch_pratap_batch_entry_handler(frm);
	},

	refresh(frm) {
		patch_pratap_batch_entry_handler(frm);
		ensure_batch_entry_button(frm);
	},
});

frappe.ui.form.on("Purchase Receipt Item", {
	add_serial_batch_bundle(frm, cdt, cdn) {
		open_batch_entry_dialog(frm, cdt, cdn);
	},
});

const pratap_batch_entry_state = {
	frm: null,
	cdn: null,
	dialog: null,
	default_pkg_qty: null,
	required_no_of_unit: null,
};

function patch_pratap_batch_entry_handler(frm) {
	if (!frm.cscript) {
		return;
	}

	frm.cscript.add_serial_batch_bundle = function () {};
}

function ensure_batch_entry_button(frm) {
	const items_field = frm.fields_dict.items;
	const grid = items_field?.grid;
	if (!grid) {
		return;
	}

	ensure_batch_entry_styles();
	bind_items_grid_selection(frm);

	items_field.$wrapper.find(".pratap-batch-entry-btn-wrapper").remove();

	const label = __("Add Batch Nos");
	const $btn = grid.add_custom_button(label, () => {
		const row = get_selected_batch_item_row(frm);
		if (!row) {
			frappe.msgprint(__("Please select an item row in the table first."));
			return;
		}
		open_batch_entry_dialog(frm, row.doctype, row.name);
	});

	$btn.detach().insertAfter(grid.wrapper.find(".grid-add-multiple-rows"));
	$btn.prop("disabled", frm.doc.docstatus !== 0);
}

function bind_items_grid_selection(frm) {
	const grid = frm.fields_dict.items?.grid;
	if (!grid || grid._pratap_batch_selection_bound) {
		return;
	}

	grid.wrapper.on("click", ".grid-row", function () {
		pratap_batch_entry_state.frm = frm;
		pratap_batch_entry_state.cdn = $(this).attr("data-name");
	});

	grid._pratap_batch_selection_bound = true;
}

function get_selected_batch_item_row(frm) {
	const grid = frm.fields_dict.items?.grid;
	if (!grid) {
		return null;
	}

	const selected = grid.get_selected_children();
	if (selected.length === 1) {
		return selected[0];
	}

	if (pratap_batch_entry_state.frm === frm && pratap_batch_entry_state.cdn) {
		const row = (frm.doc.items || []).find((item) => item.name === pratap_batch_entry_state.cdn);
		if (row) {
			return row;
		}
	}

	const item_rows = (frm.doc.items || []).filter((item) => item.item_code);
	if (item_rows.length === 1) {
		return item_rows[0];
	}

	return null;
}

async function ensure_item_batch_tracking(item_code) {
	const { message: item_meta } = await frappe.db.get_value("Item", item_code, "has_batch_no");
	if (cint(item_meta?.has_batch_no)) {
		return true;
	}

	return new Promise((resolve) => {
		frappe.confirm(
			__(
				"Item {0} is not batch tracked. Enable batch tracking on this item and continue?",
				[item_code]
			),
			async () => {
				await frappe.call({
					method: "pratap_dev.purchase_receipt_batch_entry.enable_item_batch_tracking",
					args: { item_code },
				});
				resolve(true);
			},
			() => resolve(false)
		);
	});
}

async function open_batch_entry_dialog(frm, cdt, cdn) {
	const row = locals[cdt][cdn];
	if (!row?.item_code) {
		frappe.msgprint(__("Please select an Item Code first."));
		return;
	}

	if (frm.is_new()) {
		frappe.throw(__("Please save this GRN before adding batch numbers."));
	}

	const can_continue = await ensure_item_batch_tracking(row.item_code);
	if (!can_continue) {
		return;
	}

	pratap_batch_entry_state.frm = frm;
	pratap_batch_entry_state.cdn = row.name;

	const default_pkg_qty = flt(row.custom_packing_qty) || 1;
	const required_no_of_unit = flt(row.custom_total_qty);

	const { message: context } = await frappe.call({
		method: "pratap_dev.purchase_receipt_batch_entry.get_grn_batch_entry_context",
		args: {
			purchase_receipt: frm.doc.name,
			purchase_receipt_item: row.name,
		},
		freeze: true,
		freeze_message: __("Loading batch details..."),
	});

	const batch_rows = context?.batch_rows || [];
	const initial_rows = (batch_rows.length ? batch_rows : [{}]).map((entry) =>
		prepare_batch_entry_row(entry, default_pkg_qty)
	);

	show_batch_entry_dialog(frm, row, initial_rows, default_pkg_qty, required_no_of_unit);
}

function show_batch_entry_dialog(frm, row, initial_rows, default_pkg_qty, required_no_of_unit) {
	if (pratap_batch_entry_state.dialog) {
		pratap_batch_entry_state.dialog.hide();
		pratap_batch_entry_state.dialog = null;
	}

	const is_read_only = frm.doc.docstatus !== 0;
	const bundle_label = row.serial_and_batch_bundle || __("Not set");

	const dialog = new frappe.ui.Dialog({
		title: __("Add Batch Nos"),
		size: "extra-large",
		fields: [
			{
				fieldname: "item_info",
				fieldtype: "HTML",
				options: `<div class="pratap-batch-dialog-meta">
					<div class="pratap-batch-dialog-meta-item">
						<span class="pratap-batch-dialog-label">${__("Item")}</span>
						<strong>${frappe.utils.escape_html(row.item_code || "")}</strong>
						${
							row.item_name
								? `<span class="text-muted"> — ${frappe.utils.escape_html(row.item_name)}</span>`
								: ""
						}
					</div>
					<div class="pratap-batch-dialog-meta-bundle">
						<span class="pratap-batch-dialog-label">${__("Bundle")}</span>
						<strong>${frappe.utils.escape_html(bundle_label)}</strong>
					</div>
					<div class="pratap-batch-dialog-meta-bundle">
						<span class="pratap-batch-dialog-label">${__("Required No of Unit")}</span>
						<strong>${format_batch_qty(required_no_of_unit)}</strong>
					</div>
				</div>`,
			},
			{
				fieldname: "batches",
				fieldtype: "Table",
				label: __("Batches"),
				cannot_add_rows: is_read_only,
				cannot_delete_rows: is_read_only,
				in_place_edit: !is_read_only,
				data: initial_rows,
				fields: get_batch_entry_table_fields(default_pkg_qty, is_read_only, row.item_code),
			},
		],
		primary_action_label: is_read_only ? __("Close") : __("Save Batch Nos"),
		primary_action() {
			if (is_read_only) {
				dialog.hide();
				return;
			}
			save_batch_entry_dialog(frm, row, dialog);
		},
	});

	pratap_batch_entry_state.dialog = dialog;
	pratap_batch_entry_state.default_pkg_qty = default_pkg_qty;
	pratap_batch_entry_state.required_no_of_unit = required_no_of_unit;

	dialog.show();
	style_batch_entry_dialog(dialog);

	if (!is_read_only) {
		bind_batch_entry_grid_events(dialog, default_pkg_qty, row.item_code);
	}
}

function setup_batch_link_query(grid, item_code, default_pkg_qty = 1) {
	if (!grid || !item_code) {
		return;
	}

	const batch_query = () => ({
		filters: {
			item: item_code,
		},
	});

	const batch_route_options = () => ({
		item: item_code,
	});

	const batch_field = (grid.docfields || []).find((df) => df.fieldname === "batch_no");
	if (batch_field) {
		batch_field.get_query = batch_query;
		batch_field.get_route_options_for_new_doc = batch_route_options;
	}

	const apply_query = (grid_row) => {
		const field = grid_row?.get_field?.("batch_no");
		if (field) {
			field.get_query = batch_query;
			field.df.get_query = batch_query;
			field.df.get_route_options_for_new_doc = batch_route_options;
		}
	};

	(grid.grid_rows || []).forEach(apply_query);

	if (grid._pratap_batch_query_applied) {
		return;
	}

	const original_add_new_row = grid.add_new_row.bind(grid);
	grid.add_new_row = function (...args) {
		const result = original_add_new_row(...args);
		finalize_new_batch_entry_row(grid, default_pkg_qty, item_code);
		return result;
	};

	const original_refresh = grid.refresh.bind(grid);
	grid.refresh = function (...args) {
		const result = original_refresh(...args);
		setup_batch_link_query(grid, item_code, default_pkg_qty);
		return result;
	};

	grid._pratap_batch_query_applied = true;
}

function finalize_new_batch_entry_row(grid, default_pkg_qty, item_code) {
	if (!grid) {
		return;
	}

	const default_pkg = flt(default_pkg_qty) || flt(grid._pratap_default_pkg_qty) || 1;
	const row_doc = grid.df?.data?.[grid.df.data.length - 1];
	if (row_doc) {
		finalize_batch_entry_row_doc(row_doc, default_pkg);
	}

	const grid_row = grid.grid_rows?.[grid.grid_rows.length - 1];
	if (grid_row?.doc) {
		finalize_batch_entry_row_doc(grid_row.doc, default_pkg);
		grid_row.refresh_field("custom_packing_qty");
		grid_row.refresh_field("custom_total_qty");
		grid_row.refresh_field("qty");

		if (item_code) {
			const batch_query = () => ({ filters: { item: item_code } });
			const batch_route_options = () => ({ item: item_code });
			const field = grid_row.get_field?.("batch_no");
			if (field) {
				field.get_query = batch_query;
				field.df.get_query = batch_query;
				field.df.get_route_options_for_new_doc = batch_route_options;
			}
		}
	}
}

function finalize_batch_entry_row_doc(doc, default_pkg_qty) {
	if (!flt(doc.custom_packing_qty)) {
		doc.custom_packing_qty = default_pkg_qty;
	}
	recalculate_batch_entry_row(doc, default_pkg_qty);
	return doc;
}

function style_batch_entry_dialog(dialog) {
	const grid = dialog.fields_dict.batches?.grid;
	if (!grid) {
		return;
	}

	dialog.$wrapper.find(".modal-dialog").addClass("pratap-batch-entry-dialog");
	grid.wrapper.find(".form-grid").css("overflow-x", "visible");
	grid.wrapper.find(".grid-heading-row, .grid-body .rows").css("min-width", "100%");
	grid.refresh();
}

function get_batch_entry_table_fields(default_pkg_qty, is_read_only = false, item_code = "") {
	return [
		{
			fieldname: "batch_no",
			fieldtype: "Link",
			options: "Batch",
			label: __("Batch No"),
			in_list_view: 1,
			reqd: 1,
			read_only: is_read_only,
			columns: 4,
			get_query() {
				return {
					filters: {
						item: item_code,
					},
				};
			},
			get_route_options_for_new_doc() {
				return {
					item: item_code,
				};
			},
		},
		{
			fieldname: "custom_packing_qty",
			fieldtype: "Float",
			label: __("Standard Pkg Qty"),
			in_list_view: 1,
			default: default_pkg_qty,
			read_only: 1,
			columns: 3,
		},
		{
			fieldname: "custom_total_qty",
			fieldtype: "Float",
			label: __("No of Unit"),
			in_list_view: 1,
			read_only: is_read_only,
			columns: 2,
		},
		{
			fieldname: "qty",
			fieldtype: "Float",
			label: __("Total Qty"),
			read_only: 1,
			in_list_view: 1,
			columns: 3,
		},
	];
}

function bind_batch_entry_grid_events(dialog, default_pkg_qty, item_code) {
	const grid = dialog.fields_dict.batches?.grid;
	if (!grid) {
		return;
	}

	grid._pratap_default_pkg_qty = default_pkg_qty;
	setup_batch_link_query(grid, item_code, default_pkg_qty);

	const refresh_row_fields = (grid_row) => {
		grid_row.refresh_field("custom_packing_qty");
		grid_row.refresh_field("custom_total_qty");
		grid_row.refresh_field("qty");
	};

	const on_qty_field_change = (grid_row) => {
		recalculate_batch_entry_row(grid_row.doc, default_pkg_qty);
		refresh_row_fields(grid_row);
	};

	const handler = function () {
		const $input = $(this);
		const row_name = $input.closest(".grid-row").attr("data-name");
		const grid_row = grid.grid_rows_by_docname[row_name];
		if (!grid_row) {
			return;
		}
		// Read the value straight from the input so Total Qty (Standard Pkg Qty x No of Unit)
		// updates live as the user types, instead of only on save/model-commit.
		const fieldname = $input.attr("data-fieldname");
		if (fieldname) {
			grid_row.doc[fieldname] = flt($input.val());
		}
		on_qty_field_change(grid_row);
	};

	grid.wrapper
		.off(
			"input change blur keyup",
			'input[data-fieldname="custom_packing_qty"], input[data-fieldname="custom_total_qty"]'
		)
		.on(
			"input change blur keyup",
			'input[data-fieldname="custom_packing_qty"], input[data-fieldname="custom_total_qty"]',
			handler
		);
}

function validate_batch_entry_rows(rows, required_no_of_unit) {
	if (!rows.length) {
		frappe.throw(__("Add at least one batch row."));
	}

	if (flt(required_no_of_unit) <= 0) {
		frappe.throw(__("Set No of Unit on the item row before adding batches."));
	}

	let total_no_of_unit = 0;

	for (const entry of rows) {
		const batch_no = (entry.batch_no || "").trim();
		if (!batch_no) {
			frappe.throw(__("Batch No is required for all rows."));
		}
		if (flt(entry.custom_total_qty) <= 0) {
			frappe.throw(__("No of Unit must be greater than 0 for batch {0}.", [batch_no]));
		}
		total_no_of_unit += flt(entry.custom_total_qty);
	}

	if (Math.abs(total_no_of_unit - flt(required_no_of_unit)) > 0.0001) {
		const entered_units = format_batch_qty_plain(total_no_of_unit);
		const required_units = format_batch_qty_plain(required_no_of_unit);
		const difference = format_batch_qty_plain(Math.abs(total_no_of_unit - required_no_of_unit));

		frappe.throw({
			title: __("Batch Units Mismatch"),
			message: __(
				"The total <b>No of Unit</b> across batch rows is <b>{0}</b>, but this item row requires <b>{1}</b>. The difference is <b>{2}</b>. Please adjust batch units so both totals match.",
				[entered_units, required_units, difference]
			),
		});
	}
}

function save_batch_entry_dialog(frm, row, dialog) {
	const default_pkg_qty = pratap_batch_entry_state.default_pkg_qty || flt(row.custom_packing_qty) || 1;
	const required_no_of_unit = pratap_batch_entry_state.required_no_of_unit;

	const grid = dialog.fields_dict.batches?.grid;
	const rows = (grid?.data || []).map((entry) => {
		recalculate_batch_entry_row(entry, default_pkg_qty);
		return entry;
	});

	validate_batch_entry_rows(rows, required_no_of_unit);

	dialog.hide();

	frappe.call({
		method: "pratap_dev.purchase_receipt_batch_entry.add_batches_to_grn_item",
		args: {
			purchase_receipt: frm.doc.name,
			purchase_receipt_item: row.name,
			batches: rows.map((entry) => ({
				batch_no: (entry.batch_no || "").trim(),
				standard_pkg_qty: flt(entry.custom_packing_qty) || default_pkg_qty,
				no_of_unit: flt(entry.custom_total_qty),
				total_qty: flt(entry.qty),
			})),
		},
		freeze: true,
		freeze_message: __("Updating batch details..."),
		callback(response) {
			if (response.exc) {
				return;
			}

			const result = response.message || {};
			frappe.model.set_value(row.doctype, row.name, {
				qty: result.qty,
				received_qty: result.received_qty,
				stock_qty: result.stock_qty,
				custom_packing_qty: result.custom_packing_qty,
				serial_and_batch_bundle: result.serial_and_batch_bundle,
				use_serial_batch_fields: 0,
			});

			frm.refresh_field("items");
			frappe.show_alert({
				message: __("Batch details updated"),
				indicator: "green",
			});
		},
	});
}

function prepare_batch_entry_row(row, default_pkg_qty) {
	const prepared = {
		batch_no: row.batch_no || "",
		custom_packing_qty: flt(row.standard_pkg_qty ?? row.custom_packing_qty) || default_pkg_qty,
		custom_total_qty: flt(row.no_of_unit ?? row.custom_total_qty),
		qty: flt(row.total_qty ?? row.qty),
	};
	recalculate_batch_entry_row(prepared, default_pkg_qty);
	return prepared;
}

function recalculate_batch_entry_row(row, default_pkg_qty) {
	const packing = flt(row.custom_packing_qty) || flt(default_pkg_qty) || 1;
	row.custom_packing_qty = packing;
	row.qty = packing * flt(row.custom_total_qty);
	return row;
}

function format_batch_qty(value) {
	return format_batch_qty_plain(value);
}

function format_batch_qty_plain(value) {
	const number = flt(value);
	if (!number) {
		return "0";
	}

	const formatted = flt(number, 3).toString();
	return formatted.replace(/(\.\d*?)0+$/, "$1").replace(/\.$/, "");
}

function ensure_batch_entry_styles() {
	let style = document.getElementById("pratap-batch-entry-styles");
	if (style) {
		return;
	}

	style = document.createElement("style");
	style.id = "pratap-batch-entry-styles";
	style.textContent = `
		.pratap-batch-entry-dialog {
			max-width: min(1180px, 96vw) !important;
			width: 96vw;
		}
		.pratap-batch-entry-dialog .modal-body {
			max-height: 80vh;
			overflow-y: auto;
		}
		.pratap-batch-dialog-meta {
			display: flex;
			flex-wrap: wrap;
			gap: 16px 32px;
			margin-bottom: 12px;
			padding: 10px 12px;
			border: 1px solid var(--border-color, #d1d8dd);
			border-radius: var(--border-radius, 8px);
			background: var(--subtle-fg, #f7fafc);
		}
		.pratap-batch-dialog-meta-item,
		.pratap-batch-dialog-meta-bundle {
			min-width: 220px;
		}
		.pratap-batch-dialog-label {
			display: block;
			font-size: 11px;
			text-transform: uppercase;
			letter-spacing: 0.02em;
			color: var(--text-muted, #6c7680);
			margin-bottom: 2px;
		}
		.pratap-batch-entry-dialog .form-grid {
			overflow-x: visible !important;
		}
		.pratap-batch-entry-dialog .grid-heading-row,
		.pratap-batch-entry-dialog .grid-body .rows {
			min-width: 100% !important;
		}
		.pratap-batch-entry-dialog .grid-heading-row .grid-static-col,
		.pratap-batch-entry-dialog .data-row .col {
			min-width: 0;
		}
	`;
	document.head.appendChild(style);
}

File: pratap_dev/purchase_receipt_batch_entry.py
# Copyright (c) 2026, pratap_dev contributors
# License: MIT

import json

import frappe
from frappe import _
from frappe.utils import cint, cstr, flt

from pratap_dev.purchase_receipt_batch import _get_or_create_grn_batch
from pratap_dev.purchase_receipt_batch_qc import _update_item_bundle


def _parse_batch_payload(batches):
	if isinstance(batches, str):
		batches = json.loads(batches)

	if not isinstance(batches, list):
		frappe.throw(_("Batch rows must be a list."))

	return batches


def _get_grn_item_row(grn_doc, purchase_receipt_item):
	for row in grn_doc.items:
		if row.name == purchase_receipt_item:
			return row
	return None


def _item_is_batch_tracked(item_code):
	return bool(cint(frappe.db.get_value("Item", item_code, "has_batch_no")))


def _normalize_batch_row(batch_row, default_pkg_qty):
	standard_pkg_qty = flt(batch_row.get("standard_pkg_qty")) or flt(default_pkg_qty) or 1
	no_of_unit = flt(batch_row.get("no_of_unit"))
	total_qty = flt(batch_row.get("total_qty")) or (standard_pkg_qty * no_of_unit)

	return {
		"batch_no": cstr(batch_row.get("batch_no")).strip(),
		"standard_pkg_qty": standard_pkg_qty,
		"no_of_unit": no_of_unit,
		"total_qty": total_qty,
	}


@frappe.whitelist()
def enable_item_batch_tracking(item_code):
	"""Enable batch tracking on item when user adds batches from GRN UI."""
	if not item_code or not frappe.db.exists("Item", item_code):
		frappe.throw(_("Item {0} does not exist.").format(item_code))

	frappe.db.set_value("Item", item_code, "has_batch_no", 1, update_modified=True)
	return {"item_code": item_code, "has_batch_no": 1}


@frappe.whitelist()
def get_item_batches(item_code, search_text=None, limit=50):
	"""Return batches for an item to populate batch selector in GRN UI."""
	if not item_code:
		return []

	filters = {"item": item_code}
	if search_text:
		filters["name"] = ["like", f"%{search_text}%"]

	return frappe.get_all(
		"Batch",
		filters=filters,
		fields=["name", "batch_id", "item"],
		order_by="creation desc",
		limit=cint(limit) or 50,
	)


@frappe.whitelist()
def create_grn_batch(batch_id, item_code, purchase_receipt=None):
	"""Create a batch for GRN item from the batch entry dialog."""
	batch_id = cstr(batch_id).strip()
	if not batch_id:
		frappe.throw(_("Batch No is required."))

	if not item_code or not frappe.db.exists("Item", item_code):
		frappe.throw(_("Item {0} does not exist.").format(item_code))

	if not _item_is_batch_tracked(item_code):
		enable_item_batch_tracking(item_code)

	grn = None
	if purchase_receipt and frappe.db.exists("Purchase Receipt", purchase_receipt):
		grn = frappe.get_doc("Purchase Receipt", purchase_receipt)

	purchase_receipt_doc = grn or frappe._dict(posting_date=frappe.utils.today())
	batch_no = _get_or_create_grn_batch(batch_id, item_code, purchase_receipt_doc)

	return {
		"batch_no": batch_no,
		"batch_id": batch_id,
		"item_code": item_code,
	}


@frappe.whitelist()
def get_grn_batch_entry_context(purchase_receipt, purchase_receipt_item=None):
	"""Return eligible GRN items and existing batch rows for the batch entry dialog."""
	if not purchase_receipt or not frappe.db.exists("Purchase Receipt", purchase_receipt):
		frappe.throw(_("Purchase Receipt {0} does not exist.").format(purchase_receipt))

	grn = frappe.get_doc("Purchase Receipt", purchase_receipt)
	eligible_items = []

	for row in grn.items:
		if not row.item_code or not _item_is_batch_tracked(row.item_code):
			continue

		eligible_items.append(
			{
				"name": row.name,
				"item_code": row.item_code,
				"item_name": row.item_name,
				"custom_packing_qty": flt(row.get("custom_packing_qty")) or 1,
				"custom_total_qty": flt(row.get("custom_total_qty")),
				"qty": flt(row.qty),
				"received_qty": flt(row.get("received_qty")),
				"uom": row.uom,
			}
		)

	selected_item = None
	if purchase_receipt_item:
		for grn_row in grn.items:
			if grn_row.name == purchase_receipt_item:
				selected_item = {
					"name": grn_row.name,
					"item_code": grn_row.item_code,
					"item_name": grn_row.item_name,
					"custom_packing_qty": flt(grn_row.get("custom_packing_qty")) or 1,
					"custom_total_qty": flt(grn_row.get("custom_total_qty")),
					"qty": flt(grn_row.qty),
					"received_qty": flt(grn_row.get("received_qty")),
					"uom": grn_row.uom,
					"has_batch_no": _item_is_batch_tracked(grn_row.item_code),
				}
				break
		if not selected_item:
			selected_item = next(
				(item for item in eligible_items if item["name"] == purchase_receipt_item),
				None,
			)
	elif len(eligible_items) == 1:
		selected_item = eligible_items[0]

	batch_rows = []
	item_batches = []
	if selected_item:
		from pratap_dev.pratap.doctype.pratap_quality_inspection.pratap_quality_inspection import (
			get_grn_batch_list,
		)

		item_batches = get_item_batches(selected_item["item_code"], limit=100)

		for batch in get_grn_batch_list(
			purchase_receipt,
			selected_item["item_code"],
			selected_item["name"],
		):
			batch_rows.append(
				{
					"batch_no": batch.get("batch_no"),
					"standard_pkg_qty": flt(batch.get("standard_pkg_qty"))
					or selected_item["custom_packing_qty"],
					"no_of_unit": flt(batch.get("no_of_unit")),
					"total_qty": flt(batch.get("batch_qty")),
				}
			)

	return {
		"eligible_items": eligible_items,
		"selected_item": selected_item,
		"batch_rows": batch_rows,
		"item_batches": item_batches,
	}


def _update_batch_packaging_fields(batch_no, standard_pkg_qty, batch_qty):
	"""Write Standard Pkg Qty and No of Unit (= batch qty / standard pkg qty) onto the Batch."""
	if not batch_no or not frappe.db.exists("Batch", batch_no):
		return

	meta = frappe.get_meta("Batch")
	values = {}
	pkg = flt(standard_pkg_qty)
	if meta.has_field("custom_standard_pkg_qty"):
		values["custom_standard_pkg_qty"] = pkg
	if meta.has_field("custom_no_of_unit"):
		values["custom_no_of_unit"] = (flt(batch_qty) / pkg) if pkg else 0

	if values:
		frappe.db.set_value("Batch", batch_no, values, update_modified=False)


@frappe.whitelist()
def add_batches_to_grn_item(purchase_receipt, purchase_receipt_item, batches):
	"""Create batches and Serial and Batch Bundle from custom batch entry rows."""
	batch_rows = _parse_batch_payload(batches)
	if not batch_rows:
		frappe.throw(_("Add at least one batch row."))

	if not purchase_receipt or not frappe.db.exists("Purchase Receipt", purchase_receipt):
		frappe.throw(_("Purchase Receipt {0} does not exist.").format(purchase_receipt))

	grn = frappe.get_doc("Purchase Receipt", purchase_receipt)
	if grn.docstatus != 0:
		frappe.throw(_("Purchase Receipt must be in draft to add batch numbers."))

	item_row = _get_grn_item_row(grn, purchase_receipt_item)
	if not item_row:
		frappe.throw(_("Purchase Receipt Item not found."))

	if not item_row.item_code:
		frappe.throw(_("Row {0}: Item Code is required.").format(item_row.idx))

	if not _item_is_batch_tracked(item_row.item_code):
		enable_item_batch_tracking(item_row.item_code)

	if not item_row.warehouse:
		frappe.throw(_("Row {0}: Accepted Warehouse is required.").format(item_row.idx))

	default_pkg_qty = flt(item_row.get("custom_packing_qty")) or 1
	expected_no_of_unit = flt(item_row.get("custom_total_qty"))
	if expected_no_of_unit <= 0:
		frappe.throw(_("Row {0}: Set No of Unit on the item row before adding batches.").format(item_row.idx))

	batch_map = {}
	batch_pkg_map = {}
	total_no_of_unit = 0
	standard_pkg_qty = default_pkg_qty

	for idx, raw_row in enumerate(batch_rows, start=1):
		row = _normalize_batch_row(raw_row, default_pkg_qty)
		if not row["batch_no"]:
			frappe.throw(_("Row {0}: Batch No is required.").format(idx))

		if row["no_of_unit"] <= 0:
			frappe.throw(_("Row {0}: No of Unit must be greater than 0.").format(idx))

		if row["total_qty"] <= 0:
			frappe.throw(_("Row {0}: Total Qty must be greater than 0.").format(idx))

		batch_no = _get_or_create_grn_batch(row["batch_no"], item_row.item_code, grn)
		batch_map[batch_no] = batch_map.get(batch_no, 0) + row["total_qty"]
		batch_pkg_map[batch_no] = row["standard_pkg_qty"]
		total_no_of_unit += row["no_of_unit"]
		standard_pkg_qty = row["standard_pkg_qty"]

	if abs(total_no_of_unit - expected_no_of_unit) > 0.0001:
		difference = abs(total_no_of_unit - expected_no_of_unit)
		frappe.throw(
			_(
				"Batch units mismatch: batch rows total {0} units, but this item row requires {1} units (difference: {2}). Adjust batch No of Unit values so both totals match."
			).format(total_no_of_unit, expected_no_of_unit, difference)
		)

	total_qty = sum(batch_map.values())
	_update_item_bundle(grn, item_row, batch_map, total_qty, is_rejected=False)

	# Save Standard Pkg Qty and No of Unit (= batch qty / standard pkg qty) on each Batch.
	for batch_no, batch_qty in batch_map.items():
		_update_batch_packaging_fields(
			batch_no, batch_pkg_map.get(batch_no) or standard_pkg_qty, batch_qty
		)

	item_row.qty = total_qty
	item_row.received_qty = total_qty
	item_row.stock_qty = total_qty * flt(item_row.conversion_factor or 1)
	item_row.custom_packing_qty = standard_pkg_qty
	item_row.use_serial_batch_fields = 0

	grn.save()

	return {
		"purchase_receipt_item": item_row.name,
		"item_code": item_row.item_code,
		"qty": item_row.qty,
		"received_qty": item_row.received_qty,
		"stock_qty": item_row.stock_qty,
		"custom_packing_qty": item_row.custom_packing_qty,
		"serial_and_batch_bundle": item_row.serial_and_batch_bundle,
	}
